metron-platform/metron-parsers/src/main/java/org/apache/metron/parsers/ParserRunner.java

```diff
@@ -0,0 +1,234 @@
+ */
+package org.apache.metron.parsers;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.metron.common.Constants;
+import org.apache.metron.common.configuration.FieldTransformer;
+import org.apache.metron.common.configuration.FieldValidator;
+import org.apache.metron.common.configuration.ParserConfigurations;
+import org.apache.metron.common.configuration.SensorParserConfig;
+import org.apache.metron.common.error.MetronError;
+import org.apache.metron.common.message.metadata.RawMessage;
+import org.apache.metron.common.utils.ReflectionUtils;
+import org.apache.metron.parsers.filters.Filters;
+import org.apache.metron.parsers.interfaces.MessageFilter;
+import org.apache.metron.parsers.interfaces.MessageParser;
+import org.apache.metron.parsers.topology.ParserComponent;
+import org.apache.metron.stellar.common.CachingStellarProcessor;
+import org.apache.metron.stellar.dsl.Context;
+import org.apache.metron.stellar.dsl.StellarFunctions;
+import org.json.simple.JSONObject;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class ParserRunner implements Serializable {
+
+  protected transient Consumer<ParserResult> onSuccess;
+  protected transient Consumer<MetronError> onError;
+
+  private HashSet<String> sensorTypes;
+  private Map<String, ParserComponent> sensorToParserComponentMap;
+
+  // Stellar variables
+  private transient Context stellarContext;
+
+  public ParserRunner(HashSet<String> sensorTypes) {
+    this.sensorTypes = sensorTypes;
+  }
+
+  public Map<String, ParserComponent> getSensorToParserComponentMap() {
+    return sensorToParserComponentMap;
+  }
+
+  public void setSensorToParserComponentMap(Map<String, ParserComponent> sensorToParserComponentMap) {
+    this.sensorToParserComponentMap = sensorToParserComponentMap;
+  }
+
+  public Context getStellarContext() {
+    return stellarContext;
+  }
+
+  public void setOnSuccess(Consumer<ParserResult> onSuccess) {
+    this.onSuccess = onSuccess;
+  }
+
+  public void setOnError(Consumer<MetronError> onError) {
+    this.onError = onError;
+  }
+
+  public void init(CuratorFramework client, Supplier<ParserConfigurations> parserConfigSupplier) {
+    if (parserConfigSupplier == null) {
+      throw new IllegalStateException("A parser config supplier must be set before initializing the ParserRunner.");
+    }
+    initializeStellar(client, parserConfigSupplier);
+    initializeParsers(parserConfigSupplier);
+  }
+
+  protected void initializeStellar(CuratorFramework client, Supplier<ParserConfigurations> parserConfigSupplier) {
```

Comment:
We could make the client optional here.  The only required dependency would then be ParserConfigurations.

Reply:
Agreed. A Zk connection should be optional.  Really, we might want to factor out the Stellar initialization as this most likely will need to be done slightly differently based on the execution environment.

For example, I want to use your `ParserRunner` to be able to test my parsers in the REPL.  In that case, I'm going to create my config, then test in the REPL all before it gets uploaded to Zk.  Only after I like what I have will I push the config to Zk.

Reply:
Ok sounds good.  Are you ok if I leave the Stellar stuff in there as a default that only gets called if one hasn't been provided?

Reply:
Yes, that could work.  It would make it easier to use if we have a default implementation that works in most cases.